### PR TITLE
8343039: Remove jdk.internal.misc.InternalLock and usages from java.io

### PR DESCRIPTION
**Pull request by bplb**
Uses of `InternalLock` are removed and `synchronized` is reinstated.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8343039](https://bugs.openjdk.org/browse/JDK-8343039): Remove jdk.internal.misc.InternalLock and usages from java.io (**Enhancement** - P4)

### Reviewers
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/22048/head:pull/22048` \
`$ git checkout pull/22048`

Update a local copy of the PR: \
`$ git checkout pull/22048` \
`$ git pull https://git.openjdk.org/jdk.git pull/22048/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 22048`

View PR using the GUI difftool: \
`$ git pr show -t 22048`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/22048.diff">https://git.openjdk.org/jdk/pull/22048.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/22048#issuecomment-2471658263)
</details>

### CODE REVIEW

**Comment by bplb** [comment]
Dynamic buffer resizing in `BufferedOutputStream` and `BufferedWriter` is retained.

The `jdk.internal.access` classes `JavaIOPrintStreamAccess` and `JavaIOPrintWriterAccess` will be addressed by a subsequent request.

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (ExE-Boss) [inline src/java.base/share/classes/java/io/BufferedInputStream.java:241]: This should probably still retain the lazy buffer array creation:
```suggestion
size;
        if (getClass() == BufferedInputStream.class) {
            // lazily create buffer when not subclassed
            buf = EMPTY;
        } else {
            // eagerly create buffer when subclassed
            buf = new byte[size];
        }
```

Comment (liach) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:241]: Indeed.  This and that for BufferedOutputStream are for backward compatibility with subclasses.  We cannot control when subclasses access the `buf` field after the constructor call, so laziness was impossible.

File: src/java.base/share/classes/java/io/BufferedOutputStream.java
Comment (liach) [inline src/java.base/share/classes/java/io/BufferedOutputStream.java:88]: Same remark as for BIS.

Also this initial vs max size feature was added when virtual thread was first implemented as the internal locks.  Is this implementation detail still necessary for the virtual threads?

File: src/java.base/share/classes/java/io/Writer.java
Comment (liach) [inline src/java.base/share/classes/java/io/Writer.java:279]: As annoying as this is, but https://bugs.openjdk.org/browse/JDK-8272626 didn't fix c-style arrays in method bodies.  Sad.

**Review by liach — COMMENTED** [review]
https://github.com/openjdk/jdk/tree/5212535a276a92d96ca20bdcfccfbce956febdb1 is a baseline for before virtual threads were integrated.  Note that in this period, there are other optimizations such as lazy allocation for buffer arrays or some jfr changes, so these aren't exact matches.

**Comment by AlanBateman** [comment]
I think we should remove jdk.internal.misc.InternalLock and the usages in Throwable as part of this change. I don't mind if JavaIOPrintStreamAccess is removed now or in a later PR.

File: src/java.base/share/classes/java/io/BufferedOutputStream.java
Comment (AlanBateman) [reply src/java.base/share/classes/java/io/BufferedOutputStream.java:88]: Think about 100_000 virtual threads, all blocked on sockets with input/output streams that have wrapped with buffered streams, as in BIS and BOS. This is the motivation for using a reduced initial size.

**Comment by bplb** [comment]
> I think we should remove jdk.internal.misc.InternalLock and the usages in Throwable as part of this change. I don't mind if JavaIOPrintStreamAccess is removed now or in a later PR.

I will address these items first and then the other comments, but will probably refrain from much in the way of simple cleanups, at least before the final iteration.

**Comment by bplb** [comment]
> I think we should remove jdk.internal.misc.InternalLock

`InternalLock` is also in `sun.nio.cs.Stream{De,En}coder`.

**Comment by AlanBateman** [comment]
> `InternalLock` is also in `sun.nio.cs.Stream{De,En}coder`.

Yes, these two will need to be updated. I don't mind if this is split up into several PRs but if InternalLock isn't removed by the current PR then you'll need to rename it :-)

**Comment by bplb** [comment]
> > `InternalLock` is also in `sun.nio.cs.Stream{De,En}coder`.
> 
> Yes, these two will need to be updated. I don't mind if this is split up into several PRs but if InternalLock isn't removed by the current PR then you'll need to change its title :-)

I'll go ahead and update these and remove InternalLock.

File: src/java.base/share/classes/java/io/BufferedOutputStream.java
Comment (liach) [reply src/java.base/share/classes/java/io/BufferedOutputStream.java:88]:
```suggestion
```
Seems you forgot to remove this line after restoring the if-else below.

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/BufferedInputStream.java:242]: Maybe keep "lazily create buffer when not subclassed" as the comment.

File: src/java.base/share/classes/java/io/BufferedOutputStream.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/BufferedOutputStream.java:88]: Something has got messed up here, it creates the byte[] twice now.

Comment (bplb) [reply src/java.base/share/classes/java/io/BufferedOutputStream.java:88]: Yeah, @liach caught that above: fixing.

File: src/java.base/share/classes/java/io/PrintWriter.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/PrintWriter.java:422]: I assume L422 can be deleted.

**Comment by AlanBateman** [comment]
I did a pass over the latest update, a lot of changes, it's good that Chen is reviewing too as would be too easy to introduce a bug with this refactor.

**Comment by bplb** [comment]
> [...] it's good that Chen is reviewing too as would be too easy to introduce a bug with this refactor.

Definitely on both counts.

File: src/java.base/share/classes/java/io/BufferedInputStream.java
Comment (bplb) [reply src/java.base/share/classes/java/io/BufferedInputStream.java:243]: See 9417b53.

**Comment by bplb** [comment]
f5548c1 fixes a test failure caught during CI testing.

File: test/jdk/java/lang/ProcessBuilder/Basic.java
Comment (AlanBateman) [inline test/jdk/java/lang/ProcessBuilder/Basic.java:2876]: Can you check if java.base/jdk.internal.misc can be dropped from the `@modules` in the test description?

**Comment by AlanBateman** [comment]
I've changed the PR to require at least two reviewers as I think the changes could benefit from a second reviewer.

File: src/java.base/share/classes/java/io/BufferedOutputStream.java
Comment (AlanBateman) [inline src/java.base/share/classes/java/io/BufferedOutputStream.java:89]: I think the comment is confusing, can it just say resizable when not sub-classed as it did previously?

**Review by AlanBateman — APPROVED** [review]
I did another pass over this and don't see any issues except the confusing comment in the BOS constructor.  Overall good cleanup as InternalLock was a temporary crutch to avoid pinning in the java.io area where possible.

File: src/java.base/share/classes/java/io/BufferedOutputStream.java
Comment (bplb) [reply src/java.base/share/classes/java/io/BufferedOutputStream.java]: Changed in 15bf969.

File: test/jdk/java/lang/ProcessBuilder/Basic.java
Comment (liach) [inline test/jdk/java/lang/ProcessBuilder/Basic.java:2677]: The original version in JDK 18 uses arbitrary Object and a configurable number of millis.  That said, this can stay as is as it's functionally the same as the original 18 test.

**Review by liach — APPROVED** [review]
Looks good.  Aside from the output stream smaller buffer from virtual threads, everything is pretty much a restoration to JDK18's state (aside from trivial whitespace changes)

File: src/java.base/share/classes/java/io/BufferedReader.java
Comment (bplb) [reply src/java.base/share/classes/java/io/BufferedReader.java:329]: I don't know what the indentation convention for labels is. Neither the draft style guide (https://cr.openjdk.org/~alundblad/styleguide/index-v6.html) nor the ancient code conventions from 1997 mention this.

**Comment by bplb** [comment]
> Looks good

Thanks, @liach , for the second review.

**Pull request closed without merging**
